cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -748,18 +748,77 @@
 (decl put_in_reg_mem (Value) RegMem)
 (extern constructor put_in_reg_mem put_in_reg_mem)
 
+;; Addressing modes.
+
 (type SyntheticAmode extern (enum))
 
 (decl synthetic_amode_to_reg_mem (SyntheticAmode) RegMem)
 (extern constructor synthetic_amode_to_reg_mem synthetic_amode_to_reg_mem)
 
+(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
+(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+
 (type Amode extern (enum))
 
+(decl amode_imm_reg (u32 Gpr) Amode)
+(extern constructor amode_imm_reg amode_imm_reg)
+
 (decl amode_imm_reg_reg_shift (u32 Gpr Gpr u8) Amode)
 (extern constructor amode_imm_reg_reg_shift amode_imm_reg_reg_shift)
 
-(decl amode_to_synthetic_amode (Amode) SyntheticAmode)
-(extern constructor amode_to_synthetic_amode amode_to_synthetic_amode)
+;; A helper to check if a shift amount (the `Value`) is both constant and
+;; less-than or equal to 3; this is needed since x64 can only shift addresses
+;; using two bits.
+(decl small_const_shift (u8) Value)
+(extern extractor small_const_shift small_const_shift)
+
+;; A helper to check if a constant offset value will fit into the 32-bit maximum
+;; allowed by the x64 SIB encoding. (TODO re-document, this might not be the
+;; correct explanation)
+(decl low32_will_sign_extend_to_64 (u32) Imm64)
+(extern extractor low32_will_sign_extend_to_64 low32_will_sign_extend_to_64)
+
+;; To generate an address for a memory access, we can pattern-match various CLIF
+;; sub-trees to x64's complex addressing modes (`Amode`). In pseudo-code:
+;;
+;; if address matches iadd(a, b): if either a or b: matches (ishl c with shift
+;;   amount <= 3): amode(base + offset + c >> amount) matches (iconst c where c
+;;     will sign-extend to 64 bits): amode(base + (c + offset)) matches (uextend
+;;       (iconst c) where c will sign-extend to 64 bits): amode(base + (c +
+;;     offset)) else: amode(a + b + offset >> 0) else: amode(base + offset)
+;;
+;; The rules for `to_amode` correspond to a subset of the possible addressing
+;; modes available by tweaking the SIB byte, the MOD bits, and the size of the
+;; displacement (i.e., offset). More information is available in Intel's
+;; Software Developer's Manual, volume 2, section 2.1.5, "Addressing-Mode
+;; Encoding of ModR/M and SIB Bytes."
+;;
+;; TODO should this also handle `ishl_imm`?
+(decl to_amode (MemFlags Value Offset32) Amode)
+
+;; ...matches (ishl c ...)
+(rule (to_amode flags (iadd (ishl src (small_const_shift amt)) base) offset)
+      (amode_imm_reg_reg_shift offset (put_in_gpr base) (put_in_gpr src) amt))
+(rule (to_amode flags (iadd base (ishl src (small_const_shift amt))) offset)
+      (amode_imm_reg_reg_shift offset (put_in_gpr base) (put_in_gpr src) amt))
+;; ...matches (iconst c ...)
+(rule (to_amode flags (iadd (iconst (low32_will_sign_extend_to_64 c)) base) offset)
+      (amode_imm_reg (u32_add offset c) (put_in_gpr base)))
```

Comment:
I think we need to check whether `c + offset` fits in the constant immediate, right?

If so, I think we can do this with something like (it's a hack, but):

`rule (to_amode flags (iadd (iconst (low32_will_sign_extend_to_64 c)) base) (u32_add <c offset)))`

in other words, make `u32_add` an extractor as well, and take an in-arg (use arg polarity for this) with the half of the sum we already know (`c`). For the arg-flipped version below we need to use the `u32_add` extractor inside the `low32_will_sign_extend...` instead.

(open to ideas how we might clean this up in ISLE, by the way -- aside from fitzgen's ideas around higher-order extractors, I kind of like the notion of separate constraint clauses as well, which would allow us to write something like `(given (u32_add_fits offset c))` after the LHS...)